packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,245 @@
+    ? getConsentsFromConsentGroup(content.requiredConsents.groupList)
+    : undefined;
+  const optionalConsents = content.optionalConsents
+    ? getConsentsFromConsentGroup(content.optionalConsents.groupList)
+    : undefined;
+  const consentController = useMemo(() => create({ requiredConsents, optionalConsents, cookieDomain }), []);
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const save = () => {
+    const savedData = consentController.save();
+    if (hasUserHandledAllConsents()) {
+      reRender();
+      if (content.onAllConsentsGiven) {
+        content.onAllConsentsGiven(mergeConsents());
+      }
+    }
+    return savedData;
+  };
+
+  const update: CookieConsentContextType['update'] = (key, value) => {
+    consentController.update(key, value);
+  };
+  const approveSelectedAndRequired = () => {
+    consentController.approveRequired();
+    save();
+  };
+  const approveAll: CookieConsentContextType['approveAll'] = () => {
+    consentController.approveAll();
+    save();
+  };
+  const approveRequired: CookieConsentContextType['approveRequired'] = () => {
+    Object.keys(consentController.getOptional()).forEach((optionalConsent) => {
+      update(optionalConsent, false);
+    });
+    consentController.approveRequired();
+    save();
+  };
+  const setOptional = (approved: boolean) => {
+    Object.keys(consentController.getOptional()).forEach((optionalConsent) => {
+      update(optionalConsent, approved);
+    });
+  };
+
+  const onAction: CookieConsentContextType['onAction'] = (action, consents, value) => {
+    if (action === 'approveAll') {
+      approveAll();
+    } else if (action === 'approveRequired') {
+      approveRequired();
+    } else if (action === 'approveSelectedAndRequired') {
+      approveSelectedAndRequired();
+    } else if (action === 'changeConsentGroup') {
+      consents.forEach((consent) => {
+        update(consent, value);
+      });
+    } else if (action === 'approveOptional') {
+      setOptional(true);
+    } else if (action === 'unapproveOptional') {
+      setOptional(false);
+    }
+    reRender();
+  };
+
+  const countApprovedOptional: CookieConsentContextType['countApprovedOptional'] = () => {
+    let counter = 0;
+    let approved = 0;
+    Object.values(consentController.getOptional()).forEach((isApproved) => {
+      counter += 1;
+      if (isApproved) {
+        approved += 1;
+      }
+    });
+    return approved / counter;
+  };
+
+  const areGroupConsentsApproved: CookieConsentContextType['areGroupConsentsApproved'] = (consentData) => {
+    const optionalConsentList = consentController.getOptional();
```

Comment:
Does this return all optional cookie ids where user has given consent? Or those cookie ids where user is about to give consents (checkbox has been checked)?